### PR TITLE
Convert elliptic solutions and data to use factory map

### PR DESCRIPTION
**Pull request by nilsvu**
## Proposed changes

List the available initial guesses and the analytic data for elliptic solves in `Metavariables::factory_creation`, and remove the registration code from the individual classes.

### Code review checklist

- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).
- [ ] The PR lists upgrade instructions and is labeled `bugfix` or
  `new feature` if appropriate.

### CODE REVIEW

File: src/PointwiseFunctions/InitialDataUtilities/Background.hpp
Comment (wthrowe) [inline src/PointwiseFunctions/InitialDataUtilities/Background.hpp]: Instead of adding this caveat about "static" all over the place, just don't mark the examples static.  You might need to silence a compiler warning or something.

**Review by wthrowe — CHANGES_REQUESTED** [review]

File: src/PointwiseFunctions/InitialDataUtilities/Background.hpp
Comment (nilsvu) [reply src/PointwiseFunctions/InitialDataUtilities/Background.hpp]: Pushed a fixup for this

File: tests/Unit/Elliptic/DiscontinuousGalerkin/SubdomainOperator/Test_SubdomainOperator.cpp
Comment (wthrowe) [inline tests/Unit/Elliptic/DiscontinuousGalerkin/SubdomainOperator/Test_SubdomainOperator.cpp:199]: Do you need to NOLINT twice?

**Review by wthrowe — COMMENTED** [review]
Squash, including any changes you might make because of this comment.

File: tests/Unit/Elliptic/DiscontinuousGalerkin/SubdomainOperator/Test_SubdomainOperator.cpp
Comment (nilsvu) [reply tests/Unit/Elliptic/DiscontinuousGalerkin/SubdomainOperator/Test_SubdomainOperator.cpp:199]: "Only" until we upgrade to clang-tidy 14, which adds support for `NOLINT{BEGIN,END}` :)

**Comment by nilsvu** [comment]
Squashed, thanks for reviewing @wthrowe. Please decide if you want a second review or not.

**Review by wthrowe — DISMISSED** [review]
I think we should have a second reviewer here.  While this is fairly straightforward, it does touch over 1000 lines.

**Comment by nilsvu** [comment]
@yoonso0-0 Would you be able to review this, and perhaps also #3669? It's the elliptic counterpart to your #3448.

**Comment by yoonso0-0** [comment]
@nilsleiffischer Sure. Let me first finish looking at #3648 then I'll review this and #3669  :b

**Comment by nilsvu** [comment]
Rebased on latest develop

**Comment by nilsvu** [comment]
@yoonso0-0 Please take a look at this one and https://github.com/sxs-collaboration/spectre/pull/3669 when you find the time :)

File: tests/InputFiles/Elasticity/BentBeam.yaml
Comment (yoonso0-0) [inline tests/InputFiles/Elasticity/BentBeam.yaml:21]: Q) Why do we have `&` and `*` for `iron` keyword in this input file?

File: src/PointwiseFunctions/Elasticity/ConstitutiveRelations/Tags.hpp
Comment (yoonso0-0) [inline src/PointwiseFunctions/Elasticity/ConstitutiveRelations/Tags.hpp:20]: Please document this option struct

File: src/PointwiseFunctions/InitialDataUtilities/AnalyticSolution.hpp
Comment (yoonso0-0) [inline src/PointwiseFunctions/InitialDataUtilities/AnalyticSolution.hpp:12]: Please document the namespace `elliptic::analytic_data`

File: src/PointwiseFunctions/InitialDataUtilities/AnalyticSolution.hpp
Comment (yoonso0-0) [inline src/PointwiseFunctions/InitialDataUtilities/AnalyticSolution.hpp:14]: [nitpik] representing?

File: src/PointwiseFunctions/InitialDataUtilities/InitialGuess.hpp
Comment (yoonso0-0) [inline src/PointwiseFunctions/InitialDataUtilities/InitialGuess.hpp:10]: 1. I think it'd be better move this documentation on the namespace to some other place (e.g. to `src/PointwiseFunctions/InitialDataUtilities/AnalyticSolution.hpp`)
2. Evolution systems have both 'analytic data' and 'analytic solution' which are different from each other, and I'm a bit worried that the `analytic_data` here used for naming namespace can be a little confusing. One quick solution that I could think of is just using `elliptic::initial_data` as same as evolution systems for storing abstract base classes, though I'm not sure it's a clear naming for elliptic codes.
3. Another worry is about putting these source files into `src/PointwiseFunctions/InitialDataUtilities` since evolution and elliptic source files will be mixed in a single directory, which may be fine for running code but not very pleasant for someone who tries to look at this in the future. I haven't thought about how to organize analytic data (or solution) - related base classes of elliptic systems in our refactoring procedure... :thinking: Maybe moving the files currently in `InitialDataUtilities` into a subdirectory `Evolution`?

**Review by yoonso0-0 — COMMENTED** [review]
I think I'd better continue review after resolving my comment below since it might require many lines to be changed.

File: tests/InputFiles/Elasticity/BentBeam.yaml
Comment (nilsvu) [reply tests/InputFiles/Elasticity/BentBeam.yaml:21]: This is very useful YAML syntax. See https://support.atlassian.com/bitbucket-cloud/docs/yaml-anchors/ for an explanation

File: src/PointwiseFunctions/Elasticity/ConstitutiveRelations/Tags.hpp
Comment (nilsvu) [reply src/PointwiseFunctions/Elasticity/ConstitutiveRelations/Tags.hpp:20]: I think currently very few (none?) of the option tags have a docstring, because they have a `help` string instead.

File: src/PointwiseFunctions/InitialDataUtilities/InitialGuess.hpp
Comment (nilsvu) [reply src/PointwiseFunctions/InitialDataUtilities/InitialGuess.hpp:10]: I agree with your concerns, though I don't have a great idea how to structure things better. @wthrowe what do you think?

Comment (wthrowe) [reply src/PointwiseFunctions/InitialDataUtilities/InitialGuess.hpp:10]: 1. I agree that this isn't a particularly natural place for the documentation, but I don't think `AnalyticSolution.hpp` is any better.  Some parts of the code have files entirely for namespace documentation when there's no obvious location, but that's also always felt awkward to me.
2. I don't know the standard terminology for elliptic problems.
3. I think the long-term plan is to move all the initial data and similar out of the `PointwiseFunctions` directory into the system directories.  That should result in a more natural place for these in some elliptic-specific directory.  Until then, this seems fine.

**Comment by nilsvu** [comment]
@yoonso0-0 @wthrowe I have rebased this PR to fix merge conflicts. Although I share @yoonso0-0's concerns above, I have not moved files around since I can't think of a better way to structure things right now. I suggest we get together some time to flesh out our plans for the files in `InitialDataUtilities`. For now, I think it's most important to get the refactoring to the new factory mechanism done, because upcoming PRs depend on it.

File: src/Elliptic/Utilities/GetAnalyticData.hpp
Comment (yoonso0-0) [inline src/Elliptic/Utilities/GetAnalyticData.hpp]: [nitpik] not used

Also include `DataBox.hpp` and `TMPL.hpp`

File: src/PointwiseFunctions/InitialDataUtilities/AnalyticSolution.hpp
Comment (yoonso0-0) [inline src/PointwiseFunctions/InitialDataUtilities/AnalyticSolution.hpp:14]: Q. Why an analytic solution has an initial guess? Are we just using the correct answer (analytic solution itself) for that?

File: src/PointwiseFunctions/AnalyticSolutions/Elasticity/BentBeam.cpp
Comment (yoonso0-0) [inline src/PointwiseFunctions/AnalyticSolutions/Elasticity/BentBeam.cpp]: Template parameter is no longer needed

File: src/PointwiseFunctions/AnalyticSolutions/Elasticity/Zero.hpp
Comment (yoonso0-0) [inline src/PointwiseFunctions/AnalyticSolutions/Elasticity/Zero.hpp:81]: Q. Why `pup()` function is no longer needed here?

File: src/PointwiseFunctions/AnalyticSolutions/Xcts/Factory.hpp
Comment (yoonso0-0) [inline src/PointwiseFunctions/AnalyticSolutions/Xcts/Factory.hpp:12]: Q. Can you briefly explain why a const density solution is not included here?

File: src/PointwiseFunctions/AnalyticSolutions/Xcts/Kerr.hpp
Comment (yoonso0-0) [inline src/PointwiseFunctions/AnalyticSolutions/Xcts/Kerr.hpp:171]: Just curious -- is factory conversion for `gr::Solutions` going on (perhaps at incoming PRs)?

File: src/Elliptic/Executables/Poisson/SolvePoisson.hpp
Comment (yoonso0-0) [inline src/Elliptic/Executables/Poisson/SolvePoisson.hpp:164]: Q. Why solutions are not included here for Poisson system? (though I see Lorentzian, Moustache, etc. from `AnalyticSolutions/Poisson/`)

**Review by yoonso0-0 — COMMENTED** [review]
Did my best :)  Please squash any changes

File: src/PointwiseFunctions/InitialDataUtilities/AnalyticSolution.hpp
Comment (nilsvu) [reply src/PointwiseFunctions/InitialDataUtilities/AnalyticSolution.hpp:14]: Exactly. The superclass means "this class can be used as initial guess".

File: src/PointwiseFunctions/AnalyticSolutions/Elasticity/Zero.hpp
Comment (nilsvu) [reply src/PointwiseFunctions/AnalyticSolutions/Elasticity/Zero.hpp:81]: it's provided by the superclass and doesn't need to be overridden

File: src/PointwiseFunctions/AnalyticSolutions/Xcts/Factory.hpp
Comment (nilsvu) [reply src/PointwiseFunctions/AnalyticSolutions/Xcts/Factory.hpp:12]: The `ConstantDensityStar` solution is quite old and I haven't upgraded it yet to work with the executable. It's on my list :)

File: src/PointwiseFunctions/AnalyticSolutions/Xcts/Kerr.hpp
Comment (nilsvu) [reply src/PointwiseFunctions/AnalyticSolutions/Xcts/Kerr.hpp:171]: I'm not aware of anyone working on that, but we are probably going to need it for things like `CurvedScalarWave`, where you want to factory-create the background spacetime.

File: src/Elliptic/Executables/Poisson/SolvePoisson.hpp
Comment (nilsvu) [reply src/Elliptic/Executables/Poisson/SolvePoisson.hpp:164]: I made this consistent with the other two executables 👍 It's needed anyway for simplifications we have planned.

Comment (nilsvu) [reply src/Elliptic/Executables/Poisson/SolvePoisson.hpp:164]: FYI, the reason no `AnalyticSolution` was needed in the factory classes before is, that it was assumed _all_ background classes are analytic solutions. Now I switched from `InitializeAnalyticSolution` to `InitializeOptionalAnalyticSolution`, because I'm going to delete the other one in the future to simplify things.

**Comment by nilsvu** [comment]
Thanks for your review @yoonso0-0! I pushed a fixup since I thought that might be easier for you to check. Please take a look.

File: src/Elliptic/Executables/Poisson/SolvePoisson.hpp
Comment (yoonso0-0) [reply src/Elliptic/Executables/Poisson/SolvePoisson.hpp:164]: Okay, great!

**Comment by yoonso0-0** [comment]
Looks good, please squash. Thank you!

**Review by yoonso0-0 — APPROVED** [review]

**Comment by nilsvu** [comment]
Thanks for reviewing @wthrowe and @yoonso0-0!

**Review by wthrowe — APPROVED** [review]

**Pull request merged**
